src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -701,6 +808,95 @@ public virtual int ReadByte()
+        /// </summary>
+        /// <param name="buffer">A region of memory. When this method returns, the contents of this region are replaced by the bytes read from the current stream.</param>
+        /// <param name="minimumBytes">The minimum number of bytes to read into the buffer.</param>
+        /// <param name="throwOnEndOfStream">
+        /// <see langword="true"/> to throw an exception if the end of the stream is reached before reading <paramref name="minimumBytes"/> of bytes;
+        /// <see langword="false"/> to return less than <paramref name="minimumBytes"/> when the end of the stream is reached.
+        /// The default is <see langword="true"/>.
+        /// </param>
+        /// <returns>
+        /// The total number of bytes read into the buffer. This is guaranteed to be greater than or equal to <paramref name="minimumBytes"/>
+        /// when <paramref name="throwOnEndOfStream"/> is <see langword="true"/>. This will be less than <paramref name="minimumBytes"/> when the
+        /// end of the stream is reached and <paramref name="throwOnEndOfStream"/> is <see langword="false"/>. This can be less than the number
+        /// of bytes allocated in the buffer if that many bytes are not currently available.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="minimumBytes"/> is negative, or greater than the length of <paramref name="buffer"/>.
+        /// </exception>
+        /// <exception cref="EndOfStreamException">
+        /// Thrown when <paramref name="throwOnEndOfStream"/> is <see langword="true"/> and the end of the stream is reached before reading
+        /// <paramref name="minimumBytes"/> of data.
+        /// </exception>
+        public int ReadAtLeast(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream = true)
+        {
+            ValidateReadAtLeastArguments(buffer.Length, minimumBytes);
+
+            return ReadAtLeastCore(buffer, minimumBytes, throwOnEndOfStream);
+        }
+
+        // No argument checking is done here. It is up to the caller.
+        private int ReadAtLeastCore(Span<byte> buffer, int minimumBytes, bool throwOnEndOfStream)
+        {
+            Debug.Assert(minimumBytes <= buffer.Length);
+
+            int totalRead = 0;
+            do
+            {
+                int read = Read(buffer.Slice(totalRead));
+                if (read == 0)
+                {
+                    if (throwOnEndOfStream)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    else
+                    {
+                        return totalRead;
+                    }
+                }
+
+                totalRead += read;
+            } while (totalRead < minimumBytes);
+
+            return totalRead;
+        }
+
```

Comment:
I love the fact that the new methods are not virtual. I owe a beer to everyone who pushed for that at the API review 🥇